### DISCUSSION TITLE
Custom Configsets loaded at startup of solrCloud

### DISCUSSION THREAD

**Discussion by mathieubodin** [docker-solr/docker-solr#392] [Q&A]
Hello, I look how i can start up a solrCloud (through docker-compose) fully configured with a custom *configset* to be used in production.

My purpose is to define an environnement for local development as close as possible to production. To give some context,
my idea is to manage the configset and it lifecycle outside of the solrCloud Nodes and their lifecycles. I understood i need zookeeper to spreed the configset. So i set it up, loaded into it the configset prior solrCloud startup. After the latter is booted, i can't find my custom in the list of available configsets (only _default) when i try to create a collection. Why, what am i missing ?

This process isn't well documented on solr documentation: is the configset only uploaded to zookeeper or also retrieved ?

From my understanding there is nothing like flyway or liquibase to handle migrations. Am i missing something ?

Any idea, advice would be welcomed.

**Comment by dsmiley** [comment]
Hi Mathieu,

If you're trying to have Solr start up with your configSet in place already, there is an apparently undocumented Java system property you can use when Solr starts `bootstrap_confdir` that should be the path to a configSet.  `collection.configName` will be its name, defaulting to configuration1.

> From my understanding there is nothing like flyway or liquibase to handle migrations. Am i missing something ?

Correct; there isn't.  What you can do is define new ConfigSets with modifications, create a new collection, and re-index.  You can use collection aliases so that searches continue without downtime.  Managing that is a dance clients can do but not Solr internally.  See https://solr.apache.org/guide/8_8/reindexing.html and also REINDEXCOLLECTION command.

**Reply by mathieubodin** [reply]
Thx for the quick reply

> If you're trying to have Solr start up with your configSet in place already, there is an apparently undocumented Java system property you can use when Solr starts bootstrap_confdir that should be the path to a configSet. collection.configName will be its name, defaulting to configuration1.

Correct me if i'm wrong, it means that my configset must be deployed on every node of the solrCloud cluster. In that cas what would be the point of zookeeper ?

To widen the picture, in such scenario, which application is responsible for initiate the collection (or core, it's still fuzzy to me in the solrCloud context...) ?

Concerning migration management, many thx for the tips.

**Reply by dsmiley** [reply]
> Correct me if i'm wrong, it means that my configset must be deployed on every node of the solrCloud cluster. In that cas what would be the point of zookeeper ?

Just one node would do it.  ZooKeeper isn't only used for ConfigSets :-). Also, just because you might want to bootstrap a ConfigSet in this way doesn't preclude adding others later at runtime.  Any way, this suggestion (of this property) is just one mechanism; I'm not saying it's the best path for you; only that it might be.  It's common in the bigger picture (beyond just Solr) for there to be a management service/function that initializes Solr with whatever ConfigSets are needed and possibly other configuration.

A related aside: In Solr 9, the "ConfigSetService" is pluggable, and I use this mechanism at work on a Solr fork with many custom plugins.  Even without any customizations, it's possible in Solr 9 to configure the FileSystemConfigSetService to serve the ConfigSet from disk.  That useful when you have a custom Docker image of Solr and prefer to manage a ConfigSet that way instead of ZooKeeper.

> To widen the picture, in such scenario, which application is responsible for initiate the collection

Not Solr.  Say you are the search developer/architect, you might have some simple script or it might be a service, either its own or embedded into something else.

Docker is cool but I've noticed a deficiency in the Docker ecosystem that lacks a nice way to initialize/bootstrap a service.  I've seen some Docker images with hacks/features wherein certain magic environment variables will trigger creation of something (e.g. Kafka queue or DB table or whatever).  I've seen people hack the container entrypoint to insert a script to do this stuff (yuck).  Failing that, I've seen people create a service in docker-compose (or k8s sidecar) whose sole purpose is to initialize another service and then close.  Its all hacks and I'm surprised there's been nothing better given how common the need is.
